File: go/btree/min-depth-binary-tree.go
package btree

import (
	"fmt"
	"math"
)

func minDepth(root *Node) int {
	if root == nil {
		return 0
	}
	if root.Left == nil && root.Right == nil {
		return 1
	}
	if root.Left == nil {
		return minDepth(root.Right) + 1
	}
	if root.Right == nil {
		return minDepth(root.Left) + 1
	}

	return int(math.Min(float64(minDepth(root.Left)), float64(minDepth(root.Right))) + 1)
}

func RunMinDepth() {
	tree := Tree{
		Root: &Node{
			Val: 1,
			Left: &Node{
				Val: 9,
			},
			Right: &Node{
				Val: 20,
				Left: &Node{
					Val: 15,
				},
				Right: &Node{
					Val: 7,
				},
			},
		},
	}
	fmt.Println(minDepth(tree.Root))
}

File: javascript/min-depth-binary-tree.js
'use strict';

const minDepth = (root) => {
    if (!root) return 0;
    if (!root.left) return minDepth(root.right) + 1;
    if (!root.right) return minDepth(root.left) + 1;
    if (minDepth(root.left) < minDepth(root.right)) return minDepth(root.left) + 1;
    return minDepth(root.right) + 1;
};

const root = {
    val: 3,
    left: {
        val: 9,
        left: null,
        right: null,
    },
    right: {
        val: 20,
        left: {
            val: 15,
            left: null,
            right: null,
        },
        right: {
            val: 7,
            left: null,
            right: null
        }
    }
};

console.log(minDepth(root)); // 2